docs/admin/configuration.md
---
sidebar_position: 1
---

# Configuring {{ product_name }} SMTP Mailer

When the SMTP mailer gets started, it loads the configuration from the
environment. It reads the settings in this order:

- Read environment variables which have a specific name, see section
  [Environment variables](#environment-variables).
- Load from a configuration file which defaults to `config.toml` in the current
  working directory.

## Sections in the configuration file

Functionality that can be configured through the configuration file:

- [RabbitMQ](rabbitmq.md)
- [Frontend](frontend.md)
- [Languages](languages.md)
- [SMTP server](smtp.md)
- [Support contact](support_contact.md)
- [Template builder](template_builder.md)
- [Templates](templates.md)

## Environment variables

Settings in the configuration file can be overwritten by environment variables,
nested fields are separated by two underscores `__`. The pattern looks like
this:

```sh
OPENTALK_MAIL_<field>__<nested-field>…
```

### Limitations

Some settings can not be overwritten by environment variables. This is for
example the case for entries in lists, because there is no environment variable
naming pattern that could identify the index of the entry inside the list.

### Examples

In order to set the `rabbit_mq.mail_task_queue` field, this environment variable could be used:

```sh
OPENTALK_MAIL_RABBIT_MQ__MAIL_TASK_QUEUE=opentalk_mailer
```

## Example configuration file

This file can be found in the source code distribution under `example/smtp-mailer.toml`

<!-- begin:fromfile:config/smtp-mailer.toml.md -->

```toml
# SPDX-FileCopyrightText: OpenTalk GmbH <mail@opentalk.eu>
#
# SPDX-License-Identifier: EUPL-1.2

[rabbit_mq]
# The URL to use to connect to the rabbit mq broker
url = "amqp://username:password@localhost/%2F"
# The queue to subscribe to
queue_name = "opentalk_mailer"

# The number of seconds after which the processing of a single mailer task will be aborted.
# In that case, the task will be rejected to RabbitMQ with the requeue flag on, so RabbitMQ
# will attempt to deliver it again.
#
# This should be strictly shorter than the RabbitMQ delivery acknowledge timeout
# which can be set in the RabbitMQ configuration and defaults to 30 minutes
# (1800 seconds).
#
# When set to 0, no timeout exists and the task will block infinitely if it
# doesn't finish on its own. This is the same behavior that was implemented in
# SMTP-Mailer 0.14.0 and earlier.
#task_processing_timeout_seconds = 1200

[smtp]
# SMTP Cleartext: smtp://user:pass@mailserver.example.org:1234?disable_starttls=true
# SMTP with StartTLS: smtp://user:pass@mailserver.example.org:1234
# SMTP with implicit TLS: smtps://user:pass@mailserver.example.org:1234
# Note: Encode the URL properly:
#       - Significant characters (e.g. the '@' separating the user credentials from the domain) must be written literally.
#       - Special characters within the URL parts must be paraphrased (e.g. '@' replaced with '%40')
#       Example: smtps://user%40maildomain.example.org:pass@mailserver.example.org
#smtp_server = "smtp://localhost:1025?disable_starttls=true"
# Set the From email address according to the requirements of your SMTP server.
#from_name = "OpenTalk"
#from_email = "no-reply@example.org"

[template_builder]
# Link builder templates. `base_url` and `x_id` will be replaced. Need when the frontend uses different paths.
#join_link_builder = "{base_url}/room/{room_id}"
#guest_link_builder = "{base_url}/invite/{invite_id}"
#dashboard_event_link_builder = "{base_url}/dashboard/meetings/{event_id}"

[frontend]
#base_url = "https://opentalk.example.org"
#data_protection_url = "https://opentalk.example.org/dataprotection"

[templates]
# Path to templates. These are loaded at runtime and are not compiled in.
#registered_invite = {html = "resources/templates/registered_invite.html", txt = "resources/templates/registered_invite.txt"}
#unregistered_invite = {html = "resources/templates/unregistered_invite.html", txt = "resources/templates/unregistered_invite.txt"}
#external_invite = {html = "resources/templates/external_invite.html", txt = "resources/templates/external_invite.txt"}
#registered_event_update = {html = "resources/templates/registered_event_update.html", txt = "resources/templates/registered_event_update.txt"}
#unregistered_event_update  = {html = "resources/templates/unregistered_event_update.html", txt = "resources/templates/unregistered_event_update.txt"}
#external_event_update = {html = "resources/templates/external_event_update.html", txt = "resources/templates/external_event_update.txt"}
#registered_event_cancellation = {html = "resources/templates/registered_event_cancellation.html", txt = "resources/templates/registered_event_cancellation.txt"}
#unregistered_event_cancellation = {html = "resources/templates/unregistered_event_cancellation.html", txt = "resources/templates/unregistered_event_cancellation.txt"}
#external_event_cancellation = {html = "resources/templates/external_event_cancellation.html", txt = "resources/templates/external_event_cancellation.txt"}
#registered_uninvite = {html = "resources/templates/registered_uninvite.html", txt = "resources/templates/registered_uninvite.txt"}
#unregistered_uninvite = {html = "resources/templates/unregistered_uninvite.html", txt = "resources/templates/unregistered_uninvite.txt"}
#external_uninvite = {html = "resources/templates/external_uninvite.html", txt = "resources/templates/external_uninvite.txt"}

[languages]
#default_language = "de-DE"

[monitoring]
addr = "0.0.0.0"
port = 11411

# Optional support contact information
# Adds a support contact section to the template
[support_contact]
# Support phone contact
phone = "+49123321123"
# Support mail contact
mail = "support@example.com"
```

<!-- end:fromfile:config/smtp-mailer.toml.md -->


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [0.16.0] - 2025-11-13

[0.16.0]: https://git.opentalk.dev/opentalk/backend/services/smtp-mailer/-/compare/v0.15.2...v0.16.0

### 🚀 New features

- Add from scratch container build ([!589](https://git.opentalk.dev/opentalk/backend/services/smtp-mailer/-/merge_requests/589))
- Health command ([!605](https://git.opentalk.dev/opentalk/backend/services/smtp-mailer/-/merge_requests/605), [#159](https://git.opentalk.dev/opentalk/backend/services/smtp-mailer/-/issues/159))

### 🐛 Bug fixes

- (rabbitmq) Make handling of rabbitmq tasks more robust ([!533](https://git.opentalk.dev/opentalk/backend/services/smtp-mailer/-/merge_requests/533), [#149](https://git.opentalk.dev/opentalk/backend/services/smtp-mailer/-/issues/149))
- (ci) Allow-list CVE-2025-6297 for Bookworm ([!541](https://git.opentalk.dev/opentalk/backend/services/smtp-mailer/-/merge_requests/541), [#148](https://git.opentalk.dev/opentalk/backend/services/smtp-mailer/-/issues/148))

### 🔨 Refactor

- (logging) Set log level of verbose crates separately ([!576](https://git.opentalk.dev/opentalk/backend/services/smtp-mailer/-/merge_requests/576))

### 📦 Dependencies

- (deps) Update rust crate percent-encoding to v2.3.2 ([!524](https://git.opentalk.dev/opentalk/backend/services/smtp-mailer/-/merge_requests/524))
- (deps) Update rust crate url to v2.5.6 ([!526](https://git.opentalk.dev/opentalk/backend/services/smtp-mailer/-/merge_requests/526))
- (deps) Update rust crate tokio-executor-trait to v2.2.0 ([!529](https://git.opentalk.dev/opentalk/backend/services/smtp-mailer/-/merge_requests/529))
- (deps) Lock file maintenance ([!532](https://git.opentalk.dev/opentalk/backend/services/smtp-mailer/-/merge_requests/532))
- (deps) Update rust crate fluent-templates to v0.13.1 ([!535](https://git.opentalk.dev/opentalk/backend/services/smtp-mailer/-/merge_requests/535))
- (deps) Update rust crate clap to v4.5.46 ([!536](https://git.opentalk.dev/opentalk/backend/services/smtp-mailer/-/merge_requests/536))
- (deps) Update rust crate snafu to v0.8.8 ([!537](https://git.opentalk.dev/opentalk/backend/services/smtp-mailer/-/merge_requests/537))
- (deps) Lock file maintenance ([!539](https://git.opentalk.dev/opentalk/backend/services/smtp-mailer/-/merge_requests/539))
- (deps) Update rust crate uuid to v1.18.1 ([!540](https://git.opentalk.dev/opentalk/backend/services/smtp-mailer/-/merge_requests/540))
- (deps) Update rust crate clap to v4.5.47 ([!542](https://git.opentalk.dev/opentalk/backend/services/smtp-mailer/-/merge_requests/542))
- (deps) Update rust crate log to v0.4.28 ([!544](https://git.opentalk.dev/opentalk/backend/services/smtp-mailer/-/merge_requests/544))
- (deps) Update rust crate snafu to v0.8.9 ([!543](https://git.opentalk.dev/opentalk/backend/services/smtp-mailer/-/merge_requests/543))
- (deps) Update rust crate insta to v1.43.2 ([!546](https://git.opentalk.dev/opentalk/backend/services/smtp-mailer/-/merge_requests/546))
- (deps) Update pre-commit hook fsfe/reuse-tool to v5.1.0 ([!545](https://git.opentalk.dev/opentalk/backend/services/smtp-mailer/-/merge_requests/545))
- (deps) Update pre-commit hook fsfe/reuse-tool to v5.1.1 ([!547](https://git.opentalk.dev/opentalk/backend/services/smtp-mailer/-/merge_requests/547))
- (deps) Lock file maintenance ([!548](https://git.opentalk.dev/opentalk/backend/services/smtp-mailer/-/merge_requests/548))
- (deps) Update pre-commit hook embarkstudios/cargo-deny to v0.18.5 ([!558](https://git.opentalk.dev/opentalk/backend/services/smtp-mailer/-/merge_requests/558))
- (deps) Update git.opentalk.dev:5050/opentalk/backend/containers/rust docker tag to v1.90.0 ([!556](https://git.opentalk.dev/opentalk/backend/services/smtp-mailer/-/merge_requests/556))
- (deps) Update git.opentalk.dev:5050/opentalk/tools/opentalk-ci-doc-updater docker tag to v0.2.0 ([!528](https://git.opentalk.dev/opentalk/backend/services/smtp-mailer/-/merge_requests/528))

src/settings.rs
// SPDX-FileCopyrightText: OpenTalk GmbH <mail@opentalk.eu>
//
// SPDX-License-Identifier: EUPL-1.2

use std::{
    net::IpAddr,
    path::{Path, PathBuf},
    str::FromStr,
};

use config::{Config, ConfigError, Environment, File, FileFormat};
use dirs::config_dir;
use itertools::Itertools as _;
use opentalk_types_common::users::Language;
use percent_encoding::percent_decode_str;
use serde::{Deserialize, Deserializer, Serialize, de};

#[derive(Debug, Clone, Deserialize, Default)]
pub struct Settings {
    #[serde(default)]
    pub rabbit_mq: RabbitMqConfig,

    #[serde(default)]
    pub smtp: SmtpConfig,

    #[serde(default)]
    pub templates: Templates,

    #[serde(default)]
    pub template_builder: TemplateBuilder,

    #[serde(default)]
    pub frontend: Frontend,

    #[serde(default)]
    pub languages: Languages,

    #[serde(default)]
    pub monitoring: Option<MonitoringSettings>,

    pub support_contact: Option<SupportContact>,
}

impl Settings {
    /// Creates a new Settings instance from the provided TOML file.
    /// Specific fields can be set or overwritten with environment variables (See struct level docs for more details).
    pub fn load_from_path(file_path: &Path) -> Result<Settings, ConfigError> {
        let config = Config::builder()
            .add_source(File::from(file_path).format(FileFormat::Toml))
            // MAILER is the prefix used up until v0.17 and kept for compatibility reasons
            .add_source(
                Environment::with_prefix("MAILER")
                    .prefix_separator("_")
                    .separator("__"),
            )
            // The prefix that is following the same convention as other services
            .add_source(
                Environment::with_prefix("OPENTALK_MAIL")
                    .prefix_separator("_")
                    .separator("__"),
            )
            .build()?;

        let mut warn_unknown_key = Self::warn_unused_key;
        let ignored_deserializer = serde_ignored::Deserializer::new(config, &mut warn_unknown_key);
        let settings = serde_path_to_error::deserialize(ignored_deserializer)
            .map_err(|e| ConfigError::Message(e.to_string()))?;

        log::info!("Settings loaded from \"{}\".", file_path.to_string_lossy());

        Ok(settings)
    }

    fn warn_unused_key(path: serde_ignored::Path) {
        log::warn!("Unknown configuration key: {path}");
    }

    pub fn load_from_standard_paths() -> Result<Settings, ConfigError> {
        let paths = Self::build_standard_search_paths();
        for ConfigSearchPath { path, deprecated } in &paths {
            if path.exists() {
                if *deprecated {
                    let supported_paths = paths
                        .iter()
                        .filter_map(ConfigSearchPath::display_non_deprecated)
                        .join(", ");
                    log::warn!(
                        "You're using the deprecated configuration path \"{}\", please use one of these instead: {}.",
                        path.to_string_lossy(),
                        supported_paths
                    );
                }
                return Settings::load_from_path(path);
            }
        }

        Err(ConfigError::Message(format!(
            "Couldn't find a configuration file. Searched: {}.",
            paths
                .iter()
                .map(|ConfigSearchPath { path, .. }| format!("\"{}\"", path.to_string_lossy()))
                .join(", ")
        )))
    }

    fn build_standard_search_paths() -> Vec<ConfigSearchPath> {
        let mut paths = vec![];

        paths.push(ConfigSearchPath {
            path: "config.toml".into(),
            deprecated: true,
        });

        paths.push(ConfigSearchPath {
            path: "smtp-mailer.toml".into(),
            deprecated: false,
        });

        if let Some(config_dir) = config_dir() {
            paths.push(ConfigSearchPath {
                path: config_dir.join("opentalk/smtp-mailer.toml"),
                deprecated: false,
            });
        }

        paths.push(ConfigSearchPath {
            path: "/etc/opentalk/smtp-mailer.toml".into(),
            deprecated: false,
        });

        paths
    }
}

struct ConfigSearchPath {
    path: PathBuf,
    deprecated: bool,
}

impl ConfigSearchPath {
    fn display_non_deprecated(&self) -> Option<String> {
        if self.deprecated {
            return None;
        }
        Some(format!("\"{}\"", self.path.to_string_lossy()))
    }
}

/// RabbitMQ Config containing the url to connect to
/// and the queue_name for incoming mail tasks
#[derive(Debug, Clone, Deserialize)]
pub struct RabbitMqConfig {
    #[serde(default = "rabbitmq_default_url")]
    pub url: String,
    #[serde(default = "rabbitmq_default_queue_name")]
    pub queue_name: String,
    #[serde(default = "rabbitmq_default_task_processing_timeout_seconds")]
    pub task_processing_timeout_seconds: u64,
}

impl Default for RabbitMqConfig {
    fn default() -> Self {
        Self {
            url: rabbitmq_default_url(),
            queue_name: rabbitmq_default_queue_name(),
            task_processing_timeout_seconds: rabbitmq_default_task_processing_timeout_seconds(),
        }
    }
}

fn rabbitmq_default_url() -> String {
    "amqp://guest:guest@localhost:5672".to_owned()
}

fn rabbitmq_default_queue_name() -> String {
    "opentalk_mailer".to_owned()
}

const fn rabbitmq_default_task_processing_timeout_seconds() -> u64 {
    60 * 20
}

#[derive(Debug, Clone, Deserialize)]
pub struct MonitoringSettings {
    #[serde(default = "default_monitoring_port")]
    pub port: u16,
    #[serde(default = "default_monitoring_addr")]
    pub addr: IpAddr,
}

impl Default for MonitoringSettings {
    fn default() -> Self {
        Self {
            port: default_monitoring_port(),
            addr: default_monitoring_addr(),
        }
    }
}

fn default_monitoring_port() -> u16 {
    11411
}

fn default_monitoring_addr() -> IpAddr {
    [0, 0, 0, 0].into()
}

/// A SMTP URI type
///
/// # Examples
/// SMTP Cleartext: smtp://user:pass@mailserver.example.org:1234?disable_starttls=true
/// SMTP with StartTLS: smtp://user:pass@mailserver.example.org:1234
/// SMTP with implicit TLS: smtps://user:pass@mailserver.example.org:1234
#[derive(Debug, Clone)]
pub struct SmtpUri(url::Url);
impl FromStr for SmtpUri {
    type Err = anyhow::Error;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        match url::Url::parse(s) {
            Ok(res) => match res.scheme() {
                "smtp" | "smtps" => Ok(SmtpUri(res)),
                _ => Err(anyhow::anyhow!("Invalid smtp scheme")),
            },
            Err(_) => Err(anyhow::anyhow!("Invalid smtp scheme")),
        }
    }
}

impl SmtpUri {
    pub fn disable_starttls(&self) -> bool {
        self.0
            .query_pairs()
            .any(|x| x.0 == "disable_starttls" && x.1 == "true")
    }

    pub fn port(&self) -> Option<u16> {
        self.0.port()
    }

    pub fn host(&self) -> Option<&str> {
        self.0.host_str()
    }

    pub fn scheme(&self) -> &str {
        self.0.scheme()
    }

    pub fn credentials(&self) -> anyhow::Result<(Option<String>, Option<String>)> {
        let username = self.0.username();
        let res = if username.is_empty() {
            (
                None,
                self.0
                    .password()
                    .map(percent_decode_str)
                    .map(|x| x.decode_utf8())
                    .transpose()?
                    .map(|x| x.into_owned()),
            )
        } else {
            (
                Some(percent_decode_str(username).decode_utf8()?.into_owned()),
                self.0
                    .password()
                    .map(percent_decode_str)
                    .map(|x| x.decode_utf8())
                    .transpose()?
                    .map(|x| x.into_owned()),
            )
        };

        Ok(res)
    }
}

fn smtp_uri_deserializer<'de, D>(d: D) -> Result<SmtpUri, D::Error>
where
    D: Deserializer<'de>,
{
    let string: String = Deserialize::deserialize(d)?;
    SmtpUri::from_str(&string).map_err(de::Error::custom)
}

/// SMTP Config
///
/// Configure with which SMTP server mails should be send.
#[derive(Debug, Clone, Deserialize)]
pub struct SmtpConfig {
    #[serde(
        deserialize_with = "smtp_uri_deserializer",
        default = "smtp_default_server"
    )]
    pub smtp_server: SmtpUri,
    #[serde(default = "smtp_default_from_name")]
    pub from_name: String,
    #[serde(default = "smtp_default_from_email")]
    pub from_email: String,
}

impl Default for SmtpConfig {
    fn default() -> Self {
        Self {
            smtp_server: smtp_default_server(),
            from_name: smtp_default_from_name(),
            from_email: smtp_default_from_email(),
        }
    }
}

fn smtp_default_server() -> SmtpUri {
    SmtpUri::from_str("smtp://localhost:25").unwrap()
}

fn smtp_default_from_name() -> String {
    "OpenTalk".to_string()
}

fn smtp_default_from_email() -> String {
    "no-reply@example.org".to_string()
}

/// MailTemplate Config
///
/// Paths to a set of txt and html templates for a specific mail task
#[derive(Debug, Clone, Deserialize)]
struct MailTemplate {
    txt: PathBuf,
    html: PathBuf,
}

/// Mail Templates config
///
/// Uses MailTemplate to point to sets of templates for specific mail tasks
#[derive(Debug, Clone, Deserialize)]
pub struct Templates {
    #[serde(default = "template_default_registered_invite")]
    registered_invite: MailTemplate,
    #[serde(default = "template_default_unregistered_invite")]
    unregistered_invite: MailTemplate,
    #[serde(default = "template_default_external_invite")]
    external_invite: MailTemplate,
    //
    #[serde(default = "template_default_event_update")]
    registered_event_update: MailTemplate,
    #[serde(default = "template_default_unregistered_event_update")]
    unregistered_event_update: MailTemplate,
    #[serde(default = "template_default_external_event_update")]
    external_event_update: MailTemplate,
    //
    #[serde(default = "template_default_event_cancellation")]
    registered_event_cancellation: MailTemplate,
    #[serde(default = "template_default_unregistered_event_cancellation")]
    unregistered_event_cancellation: MailTemplate,
    #[serde(default = "template_default_external_event_cancellation")]
    external_event_cancellation: MailTemplate,
    //
    #[serde(default = "template_default_uninvite")]
    registered_uninvite: MailTemplate,
    #[serde(default = "template_default_unregistered_uninvite")]
    unregistered_uninvite: MailTemplate,
    #[serde(default = "template_default_external_uninvite")]
    external_uninvite: MailTemplate,
}

impl Templates {
    pub fn iter(&self) -> impl Iterator<Item = (&Path, &str)> {
        vec![
            (self.registered_invite.txt.as_ref(), "registered_invite.txt"),
            (
                self.registered_invite.html.as_ref(),
                "registered_invite.html",
            ),
            (
                self.unregistered_invite.txt.as_ref(),
                "unregistered_invite.txt",
            ),
            (
                self.unregistered_invite.html.as_ref(),
                "unregistered_invite.html",
            ),
            (self.external_invite.txt.as_ref(), "external_invite.txt"),
            (self.external_invite.html.as_ref(), "external_invite.html"),
            //
            (
                self.registered_event_update.txt.as_ref(),
                "registered_event_update.txt",
            ),
            (
                self.registered_event_update.html.as_ref(),
                "registered_event_update.html",
            ),
            (
                self.unregistered_event_update.txt.as_ref(),
                "unregistered_event_update.txt",
            ),
            (
                self.unregistered_event_update.html.as_ref(),
                "unregistered_event_update.html",
            ),
            (
                self.external_event_update.txt.as_ref(),
                "external_event_update.txt",
            ),
            (
                self.external_event_update.html.as_ref(),
                "external_event_update.html",
            ),
            //
            (
                self.registered_event_cancellation.txt.as_ref(),
                "registered_event_cancellation.txt",
            ),
            (
                self.registered_event_cancellation.html.as_ref(),
                "registered_event_cancellation.html",
            ),
            (
                self.unregistered_event_cancellation.txt.as_ref(),
                "unregistered_event_cancellation.txt",
            ),
            (
                self.unregistered_event_cancellation.html.as_ref(),
                "unregistered_event_cancellation.html",
            ),
            (
                self.external_event_cancellation.txt.as_ref(),
                "external_event_cancellation.txt",
            ),
            (
                self.external_event_cancellation.html.as_ref(),
                "external_event_cancellation.html",
            ),
            //
            (
                self.registered_uninvite.txt.as_ref(),
                "registered_uninvite.txt",
            ),
            (
                self.registered_uninvite.html.as_ref(),
                "registered_uninvite.html",
            ),
            (
                self.unregistered_uninvite.txt.as_ref(),
                "unregistered_uninvite.txt",
            ),
            (
                self.unregistered_uninvite.html.as_ref(),
                "unregistered_uninvite.html",
            ),
            (self.external_uninvite.txt.as_ref(), "external_uninvite.txt"),
            (
                self.external_uninvite.html.as_ref(),
                "external_uninvite.html",
            ),
        ]
        .into_iter()
    }
}

impl Default for Templates {
    fn default() -> Self {
        Self {
            registered_invite: template_default_registered_invite(),
            unregistered_invite: template_default_unregistered_invite(),
            external_invite: template_default_external_invite(),
            //
            registered_event_update: template_default_event_update(),
            unregistered_event_update: template_default_unregistered_event_update(),
            external_event_update: template_default_external_event_update(),
            //
            registered_event_cancellation: template_default_event_cancellation(),
            unregistered_event_cancellation: template_default_unregistered_event_cancellation(),
            external_event_cancellation: template_default_external_event_cancellation(),
            //
            registered_uninvite: template_default_uninvite(),
            unregistered_uninvite: template_default_unregistered_uninvite(),
            external_uninvite: template_default_external_uninvite(),
        }
    }
}

fn resource_template_path(relative_path: &Path) -> PathBuf {
    std::env::var("CARGO_MANIFEST_DIR")
        .map(|s| {
            Path::new(&s)
                .join("resources/templates")
                .join(relative_path)
        })
        .unwrap_or_else(|_| Path::new("resources/templates").join(relative_path))
}

fn template_default_registered_invite() -> MailTemplate {
    MailTemplate {
        txt: resource_template_path(Path::new("registered_invite.txt")),
        html: resource_template_path(Path::new("registered_invite.html")),
    }
}

fn template_default_unregistered_invite() -> MailTemplate {
    MailTemplate {
        txt: resource_template_path(Path::new("unregistered_invite.txt")),
        html: resource_template_path(Path::new("unregistered_invite.html")),
    }
}

fn template_default_external_invite() -> MailTemplate {
    MailTemplate {
        txt: resource_template_path(Path::new("external_invite.txt")),
        html: resource_template_path(Path::new("external_invite.html")),
    }
}

fn template_default_event_update() -> MailTemplate {
    MailTemplate {
        txt: resource_template_path(Path::new("registered_event_update.txt")),
        html: resource_template_path(Path::new("registered_event_update.html")),
    }
}

fn template_default_unregistered_event_update() -> MailTemplate {
    MailTemplate {
        txt: resource_template_path(Path::new("unregistered_event_update.txt")),
        html: resource_template_path(Path::new("unregistered_event_update.html")),
    }
}

fn template_default_external_event_update() -> MailTemplate {
    MailTemplate {
        txt: resource_template_path(Path::new("external_event_update.txt")),
        html: resource_template_path(Path::new("external_event_update.html")),
    }
}

fn template_default_event_cancellation() -> MailTemplate {
    MailTemplate {
        txt: resource_template_path(Path::new("registered_event_cancellation.txt")),
        html: resource_template_path(Path::new("registered_event_cancellation.html")),
    }
}

fn template_default_unregistered_event_cancellation() -> MailTemplate {
    MailTemplate {
        txt: resource_template_path(Path::new("unregistered_event_cancellation.txt")),
        html: resource_template_path(Path::new("unregistered_event_cancellation.html")),
    }
}

fn template_default_external_event_cancellation() -> MailTemplate {
    MailTemplate {
        txt: resource_template_path(Path::new("external_event_cancellation.txt")),
        html: resource_template_path(Path::new("external_event_cancellation.html")),
    }
}

fn template_default_uninvite() -> MailTemplate {
    MailTemplate {
        txt: resource_template_path(Path::new("registered_uninvite.txt")),
        html: resource_template_path(Path::new("registered_uninvite.html")),
    }
}

fn template_default_unregistered_uninvite() -> MailTemplate {
    MailTemplate {
        txt: resource_template_path(Path::new("unregistered_uninvite.txt")),
        html: resource_template_path(Path::new("unregistered_uninvite.html")),
    }
}

fn template_default_external_uninvite() -> MailTemplate {
    MailTemplate {
        txt: resource_template_path(Path::new("external_uninvite.txt")),
        html: resource_template_path(Path::new("external_uninvite.html")),
    }
}

/// Language config
///
/// Currently only setting the default_language (a fallback) is supported
#[derive(Debug, Clone, Deserialize)]
pub struct Frontend {
    #[serde(default = "frontend_default_base_url")]
    pub base_url: String,
    #[serde(default = "frontend_default_data_protection_url")]
    pub data_protection_url: String,
}

impl Default for Frontend {
    fn default() -> Self {
        Self {
            base_url: frontend_default_base_url(),
            data_protection_url: frontend_default_data_protection_url(),
        }
    }
}

fn frontend_default_base_url() -> String {
    "https://opentalk.example.org".into()
}

fn frontend_default_data_protection_url() -> String {
    "https://opentalk.example.org/dataprotection".into()
}

/// Template Builder Config
///
/// Some information is currently not present in the mail tasks, e.g. baseURLs
/// This can be configured here.
#[derive(Debug, Clone, Deserialize)]
pub struct TemplateBuilder {
    #[serde(default = "email_default_dashboard_event_link_builder")]
    pub dashboard_event_link_builder: String,
    /// {invite_id} will be replaced
    #[serde(default = "email_default_join_link_builder")]
    pub join_link_builder: String,
    /// {invite_id} will be replaced
    #[serde(default = "email_default_guest_link_builder")]
    pub guest_link_builder: String,
}

impl Default for TemplateBuilder {
    fn default() -> Self {
        Self {
            dashboard_event_link_builder: email_default_dashboard_event_link_builder(),
            join_link_builder: email_default_join_link_builder(),
            guest_link_builder: email_default_guest_link_builder(),
        }
    }
}

fn email_default_dashboard_event_link_builder() -> String {
    "{base_url}/dashboard/meetings/{event_id}".into()
}

fn email_default_join_link_builder() -> String {
    "{base_url}/room/{room_id}".into()
}

fn email_default_guest_link_builder() -> String {
    "{base_url}/invite/{invite_id}".into()
}

/// Language config
///
/// Currently only setting the default_language (a fallback) is supported
#[derive(Debug, Clone, Deserialize)]
pub struct Languages {
    #[serde(default = "languages_default_default_language")]
    pub default_language: Language,
}

impl Default for Languages {
    fn default() -> Self {
        Self {
            default_language: languages_default_default_language(),
        }
    }
}

fn languages_default_default_language() -> Language {
    Language::from_str("en-US").expect("en-US is a valid language")
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SupportContact {
    pub phone: String,
    pub mail: String,
}

#[cfg(test)]
mod test {
    use std::{collections::BTreeMap, env};

    use serial_test::serial;

    use super::*;

    /// Test loading the settings overwritten by env vars.
    ///
    /// This test sets and reads environment variables which is inherently unsafe.
    /// Therefore it is marked as `#[serial]` so that it doesn't interfere with any other
    /// tests that might run in parallel.
    ///
    /// Once the test is finished, all variables are restored.
    #[test]
    #[serial]
    fn settings_env_vars_overwrite_config() -> Result<(), ConfigError> {
        // backup current environment variables
        let backup_vars = backup_env_variables();

        // perform the test which modifies the env variables
        let result = settings_env_vars_overwrite_config_inner();

        // restore the environment variables from the backup
        unsafe {
            restore_env_variables(backup_vars);
        }

        result
    }

    fn settings_env_vars_overwrite_config_inner() -> Result<(), ConfigError> {
        // Sanity check
        let settings = Settings::load_from_path(Path::new("./example/smtp-mailer.toml"))?;
        let support_contact = settings.support_contact.unwrap();

        assert_eq!(support_contact.phone, "+49123321123".to_string());
        assert_eq!(support_contact.mail, "support@example.com".to_string());

        // Set environment variables to overwrite default config file
        let env_support_phone = "+49777666555".to_string();
        let env_support_mail = "support@newexample.com".to_string();
        unsafe {
            env::set_var("OPENTALK_MAIL_SUPPORT_CONTACT__PHONE", &env_support_phone);
            env::set_var("OPENTALK_MAIL_SUPPORT_CONTACT__MAIL", &env_support_mail);
        }

        let settings = Settings::load_from_path(Path::new("./example/smtp-mailer.toml"))?;
        let support_contact = settings.support_contact.unwrap();

        assert_eq!(support_contact.phone, env_support_phone);
        assert_eq!(support_contact.mail, env_support_mail);

        Ok(())
    }

    fn backup_env_variables() -> BTreeMap<String, String> {
        env::vars().collect()
    }

    unsafe fn restore_env_variables(backup_vars: BTreeMap<String, String>) {
        {
            for (k, _) in env::vars() {
                if !backup_vars.contains_key(&k) {
                    unsafe { env::remove_var(&k) };
                }
            }
            for (k, v) in backup_vars {
                unsafe { env::set_var(k, v) };
            }
        }
    }
}
